specs/deposit_gas_metering.md

```diff
@@ -0,0 +1,153 @@
+single L1 block. This is to limit the amount of gas that is used on L2.
+
+This is done with a hard limit on the amount of guaranteed gas that can be
+bought. To reduce PGAs if the deposit mechanism is congested, we also implement
+an EIP-1559 style fee market with the following pseudo code:
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+GUARANTEED_GAS_LIMIT: u64, SANITY_GAS_LIMIT: u64 = load_and_unpack_storage2()
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
+    if curr_bought_gas == gas_target {
+        new_basefee := curr_basefee
+    } else if curr_bought_gas > gas_target {
+        gas_delta     := curr_bought_gas - gas_target
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        basefee_delta := max(basefee_delta, 1) # TODO: Why does 1559 have this asymmetry?
+        new_basefee   := curr_basefee + basefee_delta
+    } else {
+        gas_delta     := gas_target - curr_bought_gas
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        // Fun fact, geth doesn't let the new_basefee get below 0 and while not in the EIP spec, we should add this as well.
+        new_basefee   := curr_basefee - basefee_delta
+    }
+    curr_basefee := new_basefee
+    curr_number := block.number
+    curr_bought_gas := 0
+   
+}
+
+curr_bought_gas += required_gas
+require(curr_bought_gas <= GUARANTEED_GAS_LIMIT)
+require(curr_bought_gas <= SANITY_GAS_LIMIT)
+gas_cost = requested_gas * curr_basefee
+
+burn(gas_cost) # Via gas or eth.
+
+pack_and_store(curr_basefee, curr_number, curr_bought_gas)
+```
+
+### Rationale for Guaranteed vs Additional Gas
+
+Need to have this additional gas mechanism to enable gas refunds.  
+Guaranteed gas cannot be refunded without opening up a DOS vector.
+
+### Rationale for burning L1 Gas
+
+If we burn Eth (or collect it), we need to add the payable selector everywhere.
+Adding it everywhere is not feasible and really bad UX.
+
+## Deposits on L2
+
+### Guaranteed Gas On L2
+
+When processing L2 blocks, do the following:
+
+1. Sum up all guaranteed gas in deposits.
+2. Subtract the guaranteed gas from the gas pool.
+3. Don't later subtract guaranteed gas when processing deposits individually.
+
+The reason for subtracting out guaranteed gas at the start is to ensure that we have
+enough gas for all deposits in the presence of additional gas.
+
+### Additional Gas On L2
+
+Additional gas is only bought given the following conditions:
+
+- `gasPrice` >= the L2 gas price (however it is set)
```

Comment:
Do we need to split this into max_basefee + tip? cf. comment above on ordering by `gasPrice` 

The benefit of a max basefee is that you can specify a high one if you want to get speedy inclusion (and not overthink your parameters) and usually you won't get charged the whole amount — if there happens to be high congestion, you will simply drop out due to low tip.

The other benefit is that this is how normal gas works, and additional gas is supposed to be normal gas. This might potentially be easier to implement also (??)

Reply:
I don't think so. Sequencers have to include deposits no matter what.

Reply:
We discussed it on the call, gasPrice seems simpler, esp. this will not be used to order transactions (though there is a proposal to use something else for that, see below).